File: 2020/07/solution.rb
require_relative '../../lib/advent_helper'
helper = AdventHelper.new(script_root:__dir__)
# input = helper.send(:open_file, 'input.txt').read
# responses = input.split("\n\n").map { |r| r.split(/\s+/).map(&:chars) }
# input = helper.auto_parse
input = helper.line_separated_strings('input.txt')
# input = helper.comma_separated_strings('input.txt')



# Part 1
outside = Hash.new { |h, k| h[k] = [] }
input.each do |sentence|
  outer_bag = sentence.match(/^(.*) bags contain/)[1]
  sentence.scan(/(?:contain|,) \d* (.*?) bag/).map(&:first).each do |inner|
    outside[inner] << outer_bag
  end
end

colors = []
queue = outside["shiny gold"]

while queue.any?
  color = queue.shift
  colors << color
  outer_bags = outside[color]
  queue.concat(outer_bags)
end

a = colors.uniq.length

# Part 2
inside = Hash.new { |h, k| h[k] = [] }
input.each do |sentence|
  outer_bag = sentence.match(/^(.*) bags contain/)[1]
  inner_bags = sentence.scan(/(?:contain|,) (.*? bag)/).map(&:first)
  next if inner_bags[0] == 'no other bag'
  inner_bags.each do |string|
    number, color = string.match(/(\d+) (.*) bag/)[1..2]
    inside[outer_bag] << [number.to_i, color]
  end
end

inventory = Hash.new(0)

queue = [[1, "shiny gold"]]

while queue.length > 0
  bag = queue.shift
  c, color = bag
  inventory[color] += c
  inside[color].each do |inside_bag|
    amount, color = inside_bag
    queue << [c * amount, color]
  end
end

inventory["shiny gold"] = 0

b = inventory.values.sum



helper.output(a, b)

File: 2020/07/solution.exs
IO.write("\e[H\e[2J")
IO.puts("Running #{__ENV__.file}...")

# sample_one = "light red bags contain 1 bright white bag, 2 muted yellow bags.
# dark orange bags contain 3 bright white bags, 4 muted yellow bags.
# bright white bags contain 1 shiny gold bag.
# muted yellow bags contain 2 shiny gold bags, 9 faded blue bags.
# shiny gold bags contain 1 dark olive bag, 2 vibrant plum bags.
# dark olive bags contain 3 faded blue bags, 4 dotted black bags.
# vibrant plum bags contain 5 faded blue bags, 6 dotted black bags.
# faded blue bags contain no other bags.
# dotted black bags contain no other bags."

# sample_two = "shiny gold bags contain 2 dark red bags.
# dark red bags contain 2 dark orange bags.
# dark orange bags contain 2 dark yellow bags.
# dark yellow bags contain 2 dark green bags.
# dark green bags contain 2 dark blue bags.
# dark blue bags contain 2 dark violet bags.
# dark violet bags contain no other bags."

# rule_strings = String.split(sample_one, "\n", trim: true)

rule_strings = __DIR__ <> "/" <> "input.txt"
  |> File.read!()
  |> String.split("\n", trim: true)

rules = Enum.reduce(rule_strings, %{}, fn x, acc ->
  container_match = Regex.run(~r/\A(.*) bags contain/, x)
  key = String.to_atom(Enum.at(container_match, 1))
  cond do
    Regex.match?(~r/bags contain no other bags./, x) ->
      Map.merge(acc, %{ key => [] })
    true ->
      contained = Regex.scan(~r/(\d+) (\w+ \w+) bag/, x)
      counts = Enum.map(contained, fn match ->
        {
          Enum.at(match, 1) |> String.to_integer,
          Enum.at(match, 2) |> String.to_atom
        }
      end)
      Map.merge(acc, %{ key => counts })
  end
end)

defmodule PartOne do
  def search(rules, key) do
    rule = Map.get(rules, key)
    Enum.count(rule) != 0 &&
    Enum.any?(rule, fn {_, atom} -> atom == :"shiny gold" end) ||
    Enum.any?(rule, fn {_, atom} -> search(rules, atom) end)
  end

  def run(rules) do
    Enum.count(Map.keys(rules), fn key ->
      search(rules, key)
    end)
  end
end

defmodule PartTwo do
  def pack(rules, key) do
    rule = Map.get(rules, key)
    inside = Enum.map(rule, fn {count, atom} ->
      count * pack(rules, atom)
    end) |> Enum.sum
    inside + 1
  end

  def run(rules), do: pack(rules, :"shiny gold") - 1
end

PartOne.run(rules) |> IO.inspect
PartTwo.run(rules) |> IO.inspect
